### PR TITLE
Fix SockConect doesn't resolve domain before connecting

### LINKED ISSUE #1182: SockConnect with timeout doesn't support resolve domain
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-kvrocks/issues) and found no similar issues.

### Version

unstable

### Minimal reproduce step

1. Start Kvrocks serverA and serverB on the same machine
2. Server A becomes the slave of serverB with localhost domain

### What did you expect to see?

It's ok to use the domain(local) to connect the master

### What did you see instead?

Cannot connect to the master, related issue also can see： https://github.com/apache/incubator-kvrocks/discussions/1178

### Anything Else?

_No response_

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by git-hulk**
This closes #1182 

Currently, Kvrocks supports connecting the server with timeout or not, but the behavior is a bit different between them. The timeout version doesn't support resolving the domain before connecting to the server, so it'll be broken if use it to connect the domain.

SockConnect with timeout is the first time to use in #1172, so this bug won't affect the previous releases.

### CODE REVIEW

**Review by juancresc — DISMISSED** [review]

File: src/common/io_util.cc [hunk 1]
```diff
@@ -131,57 +97,67 @@ Status SockSetTcpKeepalive(int fd, int interval) {
...
+  hints.ai_socktype = SOCK_STREAM;
 
-    fcntl(*fd, F_SETFL, O_NONBLOCK);
-    connect(*fd, reinterpret_cast<sockaddr *>(&sin), sizeof(sin));
+  if (int rv = getaddrinfo(host.c_str(), std::to_string(port).c_str(), &hints, &servinfo); rv != 0) {
+    return {Status::NotOK, gai_strerror(rv)};
+  }
+  if (int rv = getaddrinfo(host.c_str(), std::to_string(port).c_str(), &hints, &servinfo); rv != 0) {
+    return {Status::NotOK, gai_strerror(rv)};
+  }
```
Comment (PragmaTwice) [inline src/common/io_util.cc]: Duplicated?

Comment (git-hulk) [reply src/common/io_util.cc]: Yes, it's duplicated code. I will revise this PR again and add a test case for this issue.

**Review by tisonkun — APPROVED** [review]

**Review by PragmaTwice — APPROVED** [review]

File: src/common/io_util.cc [hunk 2]
```diff
@@ -131,57 +97,66 @@ Status SockSetTcpKeepalive(int fd, int interval) {
...
-    if ((socket_arg = fcntl(*fd, F_GETFL, NULL)) < 0) {
-      return Status::FromErrno();
+    Status s = SockSetTcpKeepalive(*cfd, 120);
+    if (s.IsOK()) {
+      s = SockSetTcpNoDelay(*cfd, 1);
     }
-    socket_arg &= (~O_NONBLOCK);
-    if (fcntl(*fd, F_SETFL, socket_arg) < 0) {
-      return Status::FromErrno();
+    if (!s.IsOK()) {
```
Comment (PragmaTwice) [inline src/common/io_util.cc:145]:
```suggestion
    else {
```

File: src/common/io_util.cc [hunk 3]
```diff
@@ -131,57 +97,66 @@ Status SockSetTcpKeepalive(int fd, int interval) {
...
+      continue;
```
Comment (PragmaTwice) [inline src/common/io_util.cc:146]: continue or return error directly, which is better here?

**Review by caipengbo — APPROVED** [review]

File: src/common/io_util.cc (hunk 3 above)
Comment (git-hulk) [reply src/common/io_util.cc:146]: Looks better to continue if it has other addresses to have a try.

**Comment by git-hulk** [comment]
Thanks all, merging...

**Pull request merged**